Refactor code base, add CLI and documentation

jrobchin opened the pull request:
# Changes
- Large code refactoring, separating functionality into separate files
- Added CLI for easier interaction with backend
- Added documentation on using the CLI and running the generated code
- Added templating with Jinja2 for more flexible code generation
- Added type annotations where necessary to make code more legible
- Added unique IDs for elements to allow for easier generation and referencing
- Added classes:
  - `FigmaParser` - handles parsing Figma design data into `FigmaElement`s
  - `UserError` - raises user errors to be printed or shown in a message box
  - Encapsulating Figma data structures:
    - `FigmaFrame`
    - `FigmaElement`
    - `RectangleElement`
    - `ButtonElement`
    - `TextElement`
    - `TextEntryElement`
    - `ImageElement`

# Future Work
- Creating an installable package and uploading it to PyPi
- Convert more of Figma's design capabilities to Tkinter

piccoloser commented:
This is a really great commit! Does anything change from a user's perspective?

jrobchin commented:
@piccoloser from the user perspective, it only adds another interface. This PR should be far more attractive to devs as it'll allow for more extensibility and maintainability of the project. Of course, these benefits will eventually trickle down to the user. :)

ParthJadhav commented:
Hey @jrobchin, This is very useful and amazing contribution. Surely it will make it more attractive for devs. 

I am currently looking at code and understanding it... I'll be doing some test and updating the documentation if needed then I'll merge the PR.

Again, Thanks for this and future valuable contributions.

ParthJadhav commented:
Hey @jrobchin , I tried running the CLI but I got the following error -> 
`/Library/Frameworks/Python.framework/Versions/3.9/bin/python3: Error while finding module specification for 'tkdesigner.cli' (ModuleNotFoundError: No module named 'tkdesigner')`

It's looking for a installed module, I guess you wanted to use from tkdesigner as a directory. I tried using direct import of the files but again it's looking for installed modules.

ParthJadhav commented:
But now, The generated files has paths like this 
- build/assets/....
Which causes the code to break in case ran from inside the build folder. 

The Path's should be relative to the generated code.
- ./assets/.....

ParthJadhav commented:
I think either there are issues with the file paths or I am doing it wrong or the instructions are not complete.

jrobchin commented:
Ah right, that's an issue with the instructions, it should be `python build/gui.py`. In Python, paths are relative to where the script is run **from**.

I can make a change to get the path of the file and construct the asset paths from there as seen [here](https://stackoverflow.com/questions/918154/relative-paths-in-python). I've used this approach quite a bit and Django uses it as well for paths relative to scripts.

Once the module is turned into a package we can use the package resources API.

ParthJadhav commented:
Cool, That would work.

> Ah right, that's an issue with the instructions, it should be `python build/gui.py`. In Python, paths are relative to where the script is run **from**.

I was asking about the GUI version of Tkinter Designer. I tried almost every method to run it none worked. 

- `python3 tkdesigner/gui.py` (ModuleNotFoundError: No module named 'tkdesigner')
-  `python3 -m tkdesigner.gui`  (open "images/iconbitmap.gif": no such file or directory)
- `cd tkdesigner` 
   `python3 gui.py` (ModuleNotFoundError: No module named 'tkdesigner')
   
Out of which only second seems to work that too partially 😄

ParthJadhav commented on docs/instructions.md:
```diff
@@ -165,34 +165,55 @@ ___
-1. In Tkinter Designer, click the **Output Path** form to open a file browser
-2. Choose an output path and click **Select Folder**
-3. Press **Generate**
+1. Open TKinter Designer
```
The previous instructions to open Tkinter Designer GUI would not be valid now. We need to update it.

jvendegna replied:
Do yourself a favor and update the link to the instructions doc as well in tkinter_designer.py . It's invalid currently.

ParthJadhav replied:
Yes, I marked it in my to-do for today 😄  Discovered that in the morning.

ParthJadhav replied:
https://github.com/ParthJadhav/Tkinter-Designer/pull/73 Done !

ParthJadhav commented on tkdesigner/elements.py:
```diff
@@ -0,0 +1,176 @@
+        self.y = y
+        self.width = width
+        self.height = height
+        self.id_ = str(uuid.uuid4())[:8]
```
Can we use more generalised approach to naming ? 
Using a random string would make it difficult for elements to be identified. We could use the previous approach of naming the elements using their count.

jrobchin replied:
I don't mind changing that, but I think it should be a count across all elements as opposed to per element type. The reason being is that we then have to maintain a count for each element type. Let me know what you think.

ParthJadhav replied:
We could have a single count for all elements but it would raise confusion in the generated code. 
Eg - There would be Button1, Button and suddenly there would a Entry3. The user would think “When did I made the previous two entry 🤔”

What do you think?

jrobchin replied:
That's why I chose to do the UUIDs. We can do a count per type, not a huge deal and it would make the generated code look a bit nicer.

ParthJadhav replied:
Yup! Also if were to use UUID's then a better option could be Node ID. It's an unique ID given to each Figma element and is present in the API data. It's around 3-4 digit.

saur0x replied:
What if Figma element name is used as an identifier for Tkinter objects.
The element names should be a valid Python identifier `[a-zA-Z_]\s*`  (it can be turned into one using some regex if it's invalid) and that identifier would be used to name Canvas and Rectangle objects and also the callback functions for buttons.
Callback functions can be stored in a different module that the user can then modify to make the code functional.

ParthJadhav replied:
The thing is - Figma allows users to have elements with same name. And the main thing on which Tkinter Designer relies on is naming. 

Which means that the name of different elements would be same. And adding checks for invalid identifiers and duplicate name would make it more complex than just a count for every single element.

saur0x replied:
The identifiers can use a counter value or UUID as suffix. For example, `btn_submit_12` (`btn` specifies that it's a button even though the type is rectangle, `submit` is the name of button and `12` is unique identfier). The callback function can be named like `btn_submit_12_fn` or `btn_submit_12_command`.

It's more intuitive as the user doesn't have to keep in mind what elements translate to what Tkinter objects. Other elements can use similar naming like `txtbox_username_23` and `lbl_intro_2`. The prefix denotes the element type.

The package just has to be consistent with the naming scheme.

ParthJadhav replied:
I agree with that and I want to also add the callback function feature.

But how are we going to implement it?

jrobchin replied:
[Done](https://github.com/ParthJadhav/Tkinter-Designer/pull/67/commits/f306182a304107002ac73d0466acadec2d8e62cc)

jrobchin commented:
`python3 -m tkdesigner.gui` would be the preferred approach, I just need to fix the path issues similar to the generated code pathing issues.

ParthJadhav commented:
Cool, Thanks!

ParthJadhav commented:
Can we also add this change ? https://github.com/ParthJadhav/Tkinter-Designer/issues/68#issuecomment-877615032

jrobchin commented:
I really think we should not add changes to this PR and add new features in smaller PRs.

ParthJadhav commented:
Sure, I understand. We could add this later.

jrobchin commented:
> Cool, That would work.
> 
> > Ah right, that's an issue with the instructions, it should be `python build/gui.py`. In Python, paths are relative to where the script is run **from**.
> 
> I was asking about the GUI version of Tkinter Designer. I tried almost every method to run it none worked.
> 
> * `python3 tkdesigner/gui.py` (ModuleNotFoundError: No module named 'tkdesigner')
> * `python3 -m tkdesigner.gui`  (open "images/iconbitmap.gif": no such file or directory)
> * `cd tkdesigner`
>   `python3 gui.py` (ModuleNotFoundError: No module named 'tkdesigner')
> 
> Out of which only second seems to work that too partially

Any of below now work as of [37dc166](https://github.com/ParthJadhav/Tkinter-Designer/pull/67/commits/c98e4d2479b6227d8651b54f23eb009ec304a95b):

```bash
$ cd gui
$ python gui.py
```

`$ python gui/gui.py`

`$ python -m gui.gui`

jrobchin commented:
> Ah right, that's an issue with the instructions, it should be `python build/gui.py`. In Python, paths are relative to where the script is run **from**.
> 
> I can make a change to get the path of the file and construct the asset paths from there as seen [here](https://stackoverflow.com/questions/918154/relative-paths-in-python). I've used this approach quite a bit and Django uses it as well for paths relative to scripts.
> 
> Once the module is turned into a package we can use the package resources API.

This has been implemented as of [9e2d5d](https://github.com/ParthJadhav/Tkinter-Designer/pull/67/commits/9e2d5d34b81614976f388ad319a892c537d39767)

jrobchin commented:
Should be ready for another review :+1:

ParthJadhav commented on tkdesigner/elements.py:
```diff
@@ -0,0 +1,172 @@
+    image=button_image_{self.id_},
+    borderwidth=0,
+    highlightthickness=0,
+    command=lambda _: print("button_{self.id_} clicked"),
```
lambda is requiring 1 argument which is not being passed and also not needed. It's throwing error. 

Fix -> 
`command=lambda: print("button_{self.id_} clicked")`

ParthJadhav replied:
Fixed in https://github.com/ParthJadhav/Tkinter-Designer/pull/67/commits/f10414b845e1fb65eac7ee714ea97696bb720124

ParthJadhav commented on tkdesigner/cli.py:
```diff
@@ -0,0 +1,70 @@
+    parser.add_argument("file_url", type=str, help="File url of the Figma design.")
+    parser.add_argument("token", type=str, help="Figma token.")
+    parser.add_argument("-o", "--output", type=str, default="build/", help="Folder to output code and image assets to. Defaults to build/.")
+    parser.add_argument("-f", "--force", action="store_true", help="If this flag is passed in, the output directory given will be overwritten if it exists.")
```
This does not seem to work. Even if the directory exists it overrides it without any prompt even if -f is not passed.

saur0x replied:
How is one supposed to execute the cli.py module?
I get `ModuleNotFoundError: No module named 'tkdesigner'` when I execute it from inside the `tkdesigner` directory or from its parent directory `Tkinter-Designer`.

ParthJadhav commented:
CLI version now works great ! And also now I can use the GUI version but the files are not generated through GUI version. 

Also the numbering thing is not implemented I guess. I see the code for it but It does not work. - 
<img width="547" alt="Screenshot 2021-07-11 at 10 22 38 AM" src="https://user-images.githubusercontent.com/42001064/125183162-ec42ae80-e231-11eb-99bb-dfc809244c84.png">

ParthJadhav commented:
Changes :-

Bug Fixes -
1. Lambda function of Button's generated code was requiring an extra argument causing the button to crash because that argument was not passed. Removed that argument to fix the bug.

Naming Changes :-
Images were named using UUID while their respective code were named using a counter. Changed both to be named using counter to maintain uniformity.

Found Bugs :-
1. The code generation is not working through the GUI version.
2. -f --force argument is not working. The code overwrites without the argument.

neilbaldwin commented:
This is a great progression! :)

One question. @ParthJadhav how come you removed the text anchor? Have a better fix?

ParthJadhav commented:
> One question. @ParthJadhav how come you removed the text anchor? Have a better fix?

For two reasons - 

1. This PR contains changes related to refactoring and improving the project architecture. So a improvement would be out of scope.
2. If I make that change then you won't get the credit for it. I want you to make that same change once this PR is merged and then create a PR for that.

neilbaldwin commented:
> > One question. @ParthJadhav how come you removed the text anchor? Have a better fix?
> 
> For two reasons -
> 
>     1. This PR contains changes related to refactoring and improving the project architecture. So a improvement would be out of scope.
> 
>     2. If I make that change then you won't get the credit for it. I want you to make that same change once this PR is merged and then create a PR for that.

Haha, I don't need a credit but I appreciate the consideration! Thank you 😊

neilbaldwin commented:
Just tried the cli and it works really well. Amazing project you guys ✊🏻🙏🏻

ParthJadhav commented:
💥😃

ParthJadhav commented:
💥😃

jrobchin commented:
> Changes :-
> 
> Bug Fixes -
> 
> 1. Lambda function of Button's generated code was requiring an extra argument causing the button to crash because that argument was not passed. Removed that argument to fix the bug.
> 
> Naming Changes :-
> Images were named using UUID while their respective code were named using a counter. Changed both to be named using counter to maintain uniformity.
> 
> Found Bugs :-
> 
> 1. The code generation is not working through the GUI version.
> 2. -f --force argument is not working. The code overwrites without the argument.

Both bugs fixed in [cdad8b4](https://github.com/ParthJadhav/Tkinter-Designer/pull/67/commits/cdad8b42fb55f8a75ae1c18d6b0062eb79a96436) and [9c6d5aa](https://github.com/ParthJadhav/Tkinter-Designer/pull/67/commits/9c6d5aa91f8d849fff622b2e1140d2a5bbc2a69c) :+1:

jrobchin commented:
> Naming Changes :-
> Images were named using UUID while their respective code were named using a counter. Changed both to be named using counter to maintain uniformity.

Good catch :)

ParthJadhav commented:
> Both bugs fixed in [cdad8b4](https://github.com/ParthJadhav/Tkinter-Designer/pull/67/commits/cdad8b42fb55f8a75ae1c18d6b0062eb79a96436) and [9c6d5aa](https://github.com/ParthJadhav/Tkinter-Designer/pull/67/commits/9c6d5aa91f8d849fff622b2e1140d2a5bbc2a69c) 👍

Awesome ! I tested both GUI and CLI and they work really well. I can merge it after I add the instructions. 

Just one BIG THING -> I lost my Desktop files because of the -f flag 😕 I didn't read the code properly of the -f flag. Can we remove that feature overall or make it better so it does not delete the files ? 
_The files were not much imp though_

jrobchin commented:
Would you rather add a check that confirms with the user or remove it all together?

ParthJadhav commented:
Look at this -> this is a previous implementation of the output code.

Where generated directory is - Output Path + "/generated_files" this way it didn't checked the parent directory to be empty. And only replaced the generated directory.


        `try:
            os.mkdir(generated_dir)

        except FileExistsError:
            messagebox.showinfo("File Exists",
                                "Existing Files will be overwritten.")`

jrobchin commented:
I don't like the idea of overwriting by default. The output path can be passed in explicitly with `-o/--output` as well. Force is the way to go IMHO. If you pass in the `-f` flag then you should expect files to be overwritten, don't you think?

ParthJadhav commented:
Yes, I totally agree about that. But I shared that code for a different reason.  

You see the current implementation of -force is :- 

```
if args.force:
            logging.info(f"Overwritting exisiting output directory")
            shutil.rmtree(args.output)
            create_dirs(args.output, asset_path)
```

Which deletes the directory tree if it's not empty. Which would cause problems. I am saying that we can just overwrite the files without deleting the directory tree.

ParthJadhav commented:
In case of GUI version it defaults to overwrite files and that can be fixed by a confirmation prompt if the directory exists in future versions. But CLI must be fixed now as it's a new addition and could cause problems to users.

jrobchin commented:
> Yes, I totally agree about that. But I shared that code for a different reason.  
> 
> 
> 
> You see the current implementation of -force is :- 
> 
> 
> 
> ```
> 
> if args.force:
> 
>             logging.info(f"Overwritting exisiting output directory")
> 
>             shutil.rmtree(args.output)
> 
>             create_dirs(args.output, asset_path)
> 
> ```
> 
> 
> 
> Which deletes the directory tree if it's not empty. Which would cause problems. I am saying that we can just overwrite the files without deleting the directory tree.

I don't think this is right either for a few reasons. If we don't clear the old files: 
- The assets folder will be filled with old assets
- If something about the generated code changes in the project and a user updates, this could cause issues with newly generated code living with previously generated code

In my opinion the build folder should always be clean before files are written to it. The only way to do that is to expect the user to choose a non-existent file path or allow them to force delete it like we have it.

ParthJadhav commented:
Yes, You are right ! How about this :- 
A new directory with a different name would be created if a previous one exists.  Eg -
- Build_1
- Build_2

neilbaldwin commented:
Personally I don’t think generating multiple output folders (build1, build2 etc) is the right way.

I like the force flag. There’s no ambiguity in my mind if you choose that option. Maybe an insurance could be to somehow fingerprint files created by output process and delete the files in the output folder one by one checking for the fingerprint. If there is a file that has no fingerprint then it might be best to assume that either the use has made a mistake in selecting the output folder or that they’ve inadvertently put a folder in the output path that would get erased with the force flag?

ParthJadhav commented:
Yes, Exactly. I don't have any problem with the force flag if it doesn't erase the entire directory if it's not empty.

jrobchin commented:
Okay that's not a bad idea. I could write some kind of file that keeps track of all the written files so we know which ones we created. I'll try that.

ParthJadhav commented:
Cool !

saur0x commented:
I don't think the library or the GUI should be responsible for emptying the specified output directory. Deleting filesystem content is a sensitive task and should thus be the responsibility of the user.

The behavior of `--force` flag can be as follow if the directory is not empty:
- If the flag is off, then the program should prompt whether to continue writing to the directory or to abort.
- If the flag is on, then the generated files are written to the directory regardless of whether the directory is empty or not.

ParthJadhav commented:
Hey @jrobchin , You have already made outstanding changes through this PR and Thanks for that. But this PR is yet to merge because of that sensitive bug with the CLI version which needs to be fixed before the PR is merged. There had been different opinions on what to do with the -f or --force flag some of them are :-

1️⃣ By @saur0x 
> I don't think the library or the GUI should be responsible for emptying the specified output directory. Deleting filesystem content is a sensitive task and should thus be the responsibility of the user.
> 
> The behavior of --force flag can be as follow if the directory is not empty:
> 
> If the flag is off, then the program should prompt whether to continue writing to the directory or to abort.
> If the flag is on, then the generated files are written to the directory regardless of whether the directory is empty or not.

2️⃣ By @neilbaldwin & @jrobchin 
> I like the force flag. There’s no ambiguity in my mind if you choose that option. Maybe an insurance could be to somehow fingerprint files created by output process and delete the files in the output folder one by one checking for the fingerprint. If there is a file that has no fingerprint then it might be best to assume that either the user has made a mistake in selecting the output folder or that they’ve inadvertently put a file/folder in the output path that would get erased with the force flag?

3️⃣ By @ParthJadhav 
> We can just overwrite the files without deleting the directory tree Or Create a new directory with a new name.

I would ask all the contributors who participated in this PR to vote for the best course of action. 
@jrobchin  , @piccoloser , @ParthJadhav , @neilbaldwin & @saur0x 

As for me I think all are good options but I am on side of @saur0x  method.

@jrobchin you can always transfer your workload to me if you are busy or if you feel tried. After all is a tedious task to work on a single thing for a long time with lot of repetition 😄

piccoloser commented:
Personally, I agree with the first idea in which the user is given a prompt before the program does anything, and if the program is run with the `-f|--force` flag, the prompt assumes that they want to overwrite the files.

This method's actually fairly common, where the user will be asked to provide a `Y/n` response for sensitive operations, but the `-f` flag just inputs `Y` automatically.

Example (without `-f`):
```
The output folder provided already exists and is not empty.
Continuing with this operation will overwrite the files in this
location. Would you still like to proceed (Y/n)?
```
This makes it a manual assertion that you _want_ to overwrite the files, seeing as you had to go out of your way to use the "force" flag, and otherwise you're given a choice before the program proceeds.

jrobchin commented:
I won't be able to make these changes for a few days if someone wants to take a shot.

saur0x commented:
I have a few changes over @jrobchin 's fork and I am confused about how I should proceed.
Which base repository should I make the PR in? Here is the repository https://github.com/saur0x/Tkinter-Designer/tree/figma_api_refactor .

ParthJadhav commented:
> I won't be able to make these changes for a few days if someone wants to take a shot.

That's fine, Me and @saur0x will be working on it.

ParthJadhav commented:
Hey @piccoloser , If you have time can you review the new instructions in this commit -> https://github.com/ParthJadhav/Tkinter-Designer/pull/67/commits/ffa93ec25526481ef5883db799d762e2d925cbc8 ?

jrobchin commented:
Looks like we might be ready to merge soon!

neilbaldwin commented:
Exciting!

piccoloser commented:
@ParthJadhav Sure thing, sorry for the late response!

ParthJadhav commented:
> @ParthJadhav Sure thing, sorry for the late response!

No Problem, You must be gaming on your new PC 😄

piccoloser commented:
You know it lol

saur0x commented:
I have a very basic GitHub question. Why do some of my commits show my profile picture while others don't? Is it a caching issue in my web browser?

piccoloser commented:
@ParthJadhav I left a few comments for grammatical and clarification stuff on that commit. Everything looks alright as long as the renamed links work!

I refrained from commenting on the commands, but in Windows you don't have to capitalize anything in the command line, so you could change `cd Tkinter-Designer` to `cd tkinter-designer`, but I'm not sure if that would still work on other OSs

ParthJadhav commented:
Hey @piccoloser , Thanks for the pointers. I have fixed them.

ParthJadhav commented:
### Awesome work everyone 🔥 
This Pull Request would make Tkinter Designer a better repository for dev's to work with. 

Thank you for the refactoring @jrobchin & @saur0x 👏
Thanks you for suggestions @piccoloser 👏

**Merging the Pull Request**   🎉

jrobchin commented:
Awesome, thanks everyone 😁
